0001-dinerojs-v2.md

```diff
@@ -0,0 +1,188 @@
+
+```js
+import Dinero from "dinero.js";
+import { IQD } from "@dinero.js/currencies";
+
+// `precision` is set to `2` (default)
+const d1 = Dinero({ amount: 1000, currency: "IQD" });
+
+// `precision` is automatically set to `3`
+const d2 = Dinero({ amount: 1000, currency: IQD });
+```
+
+`Currency` objects could also embark the currency's symbol, allowing the library to fully own the currency formatting within the [`toFormat`](https://sarahdayan.github.io/dinero.js/module-Dinero.html#~toFormat) method and liberate itself from its dependency on the I18n API.
+
+```js
+import Dinero from "dinero.js";
+import { EUR } from "@dinero.js/currencies";
+
+const d1 = Dinero({ amount: 100, currency: "EUR" });
+const d2 = Dinero({ amount: 100, currency: EUR });
+
+d1.toFormat("$0,0.00"); // EUR100.00
+d2.toFormat("$0,0.00"); // €100.00
+```
+
+`Dinero` objects created with a string `currency` instead of a `Currency` object don't have any information on the currency symbol. Consequently, they would output the currency code in full instead of the symbol.
+
+The I18n API could still be used on Dinero objects as with v1, by chaining it to a `toRoundedUnit` method call: `Dinero().toRoundedUnit(digits, roundingMode).toLocaleString(locale, options)`.
+
+Users would also be able to create their own, custom currencies, breaking out of the ISO 4217 limit.
+
+```js
+const FOO = {
+  name: "foo",
+  symbol: "F",
+  code: "FOO",
+  exponent: 4
+};
+```
+
+TypeScript would further facilitate it, as they would have access to a `Currency` type.
+
+```ts
+const FOO: Currency = {
+  name: "foo",
+  symbol: "F",
+  code: "FOO",
+  exponent: 4
+};
+```
+
+#### Helpers
+
+Dinero v1 constrains users always to create `Dinero` objects the same way, using the main `Dinero` factory. It is limiting for several reasons:
+
+- It requires passing amounts as integers, which requires that users perform manual conversion when their initial data isn't in the right format.
+- It relies on globals to facilitate the creation of many objects with the same currency. It creates a limit to a single currency at a time and makes changing the default imperative.
+
+Helpers solve that problem by letting the user pass information in various formats, or use currency factories as shortcuts.
+
+```js
+import { fromFloat, fromFormat, DineroUSD } from "@dinero.js/helpers";
+
+// returns a Dinero object with `amount` equal to 49999
+// and `precision` equal to `2`
+fromFloat({ amount: 499.99 });
+
+// returns a Dinero object with `amount` equal to 15000,
+// `precision` equal to `2` and `currency` equal to `USD`
+// (using the `Currency` service under the hood)
+fromFormat("$150", "$0");
+
+// returns a Dinero object with `amount` equal to 4500
+// and precision equal to `2` (using the `Currency` service under the hood)
+DineroUSD(45);
```

Comment:
Is this exposed for every currency, or just because USD is the default? 

I think it would be interesting to remove the default if that makes code simpler

Reply:
I think it could be interesting to have one per currency: `DineroUSD`, `DineroEUR`, `DineroIQD`, etc.

That's what MoneyPHP does, they have static methods on their `MoneyFactory` class to rapidly create objects. It's quite nice because you don't have to care about passing the currency nor thinking about exponents or converting your amount into sub-units.

Reply:
Given how light passing the currency in is, this seems not particularly valuable, but don't feel strongly about it

Reply:
@osdiab It's purely utilitarian. I may or may not add it in the end, one can do it easily by importing the `Currency` object.

Reply:
I would consider retaining the existing api (to some extent) but having, e.g `currency = 'USD'`  be backed by an "object" (object as map), where the lookup key is `USD` and it maps to the CurrencyUSD object, which is then returned or delegated to.

Reply:
@xenoterracide The more I'm thinking about the helpers, the less I think I'm going to add the factories for currency-flavored Dinero objects (like `DineroUSD`, `DineroEUR`, etc.)

The main reason is that if I do that, I need to provide implementations of currency objects, which is problematic because if I make the slightest mistake in a currency declaration (or if someone disagrees), we're locked in a situation where the smallest change means breaking the API. The thing is, information about currency isn't always straightforward, and often up to interpretation. In some cases, such as Arabic currencies, I'm not even qualified to make a decision because I simply do not understand the alphabet. So it's much safer for me to provide the interface and let people implement the way they want. This way, everyone does what makes sense to them, no need to vainly accommodate everyone.

Regarding your proposition (keeping string-based `currency` to use to find definitions in a dictionary, if I got your message right), it would therefore not be possible, because it implies that Dinero.js provides implementations out of the box, and as described above, it creates issues. Also, doing something like that would imply loading all the currencies internally so they can be matched from the passed string. You definitely want to avoid that from a performance standpoint.

Reply:
@sarahdayan I would argue that while we can't be certain that a currency won't change over time, or that we'll get it right the first time, we do have access to semver and breaking version numbers.

In my case I'm using major currencies only, and having to specify precision everywhere is annoying and error prone. And having to reimplement this lookup in every project is silly too if you ask me. The library can know how USD works and share that knowledge for everyone.

I would vastly prefer a `currency = 'USD'` style API, and if any currencies change, we use semver and release Dinero v3, marking it as a breaking change in the release notes.

Reply:
I think if you have a "CurrencyRegistry", and provide default implementations which can be added to the registry if users want to operate with that currency you are in the best of all worlds. Users are free to register new currencies, or replace currencies in the registry with their preferences, but you can provide sane defaults for many currencies which users are free to use or not as they see fit.

Reply:
So, in the end there will be a separate package called `@dinero.js/currencies` with implementation for major currencies (all ISO 4217, many crypto-currencies). You can see what it looks like on the [v2 release branch](https://github.com/dinerojs/dinero.js/tree/release/v2/packages/currencies/src).

The objects are minimal, which means it's unlikely to get them wrong. If one changes for economic reasons, then this package will bump a major.